contracts/liquidity/factory/src/execute.rs

```diff
@@ -835,6 +839,102 @@ pub fn receive_cw20(
                 &mut deps, env, info, sender, asset_in, amount_in, timeout, recipient,
             )
         }
+        FactoryCw20HookMsg::EuclidReceive(euclid_receive) => {
+            receive_euclid_cw20(deps, env, info, sender, cw20_msg.amount, euclid_receive)
+        }
+    }
+}
+
+pub fn receive_euclid_native(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    euclid_receive: EuclidReceive,
+) -> Result<Response, ContractError> {
+    match from_json::<FactoryEuclidReceiveHook>(euclid_receive.data.clone())? {
+        FactoryEuclidReceiveHook::Swap {
+            sender,
+            asset_in,
+            asset_out,
+            min_amount_out,
+            swaps,
+            timeout,
+            cross_chain_addresses,
+            partner_fee,
+        } => {
+            ensure!(
+                asset_in.token_type.is_native(),
+                ContractError::InvalidAsset {
+                    asset: asset_in.token.to_string(),
+                }
+            );
+            let swap_msg = ExecuteSwapRequest {
+                sender: sender.clone(),
+                asset_in,
+                amount_in: Uint128::zero(),
+                asset_out,
+                min_amount_out,
+                swaps,
+                timeout,
+                cross_chain_addresses,
+                partner_fee,
+            };
+            let response = crate::contract::execute(
+                deps,
+                env,
+                info,
+                ExecuteMsg::ExecuteSwapRequest(swap_msg),
+            )?;
+            let event = simple_event()
+                .add_attribute("meta", euclid_receive.meta.unwrap_or("no_meta".to_string()));
+            Ok(response.add_event(event))
+        }
+    }
+}
+
+pub fn receive_euclid_cw20(
```

Comment:
It's still possible to call `Swap` directly from `receive_cw20`, is this intentional?